src/libs/sfdk/virtualmachine.h

```diff
@@ -0,0 +1,211 @@
+** conditions contained in a signed written agreement between you and Digia.
+**
+****************************************************************************/
+
+#pragma once
+
+#include "sfdkglobal.h"
+#include "asynchronous.h"
+
+#include <QUrl>
+
+#include <memory>
+
+namespace QSsh {
+class SshConnectionParameters;
+}
+
+namespace Sfdk {
+
+class SFDK_EXPORT VirtualMachineDescriptor
+{
+public:
+    QUrl uri;
+    QString displayType;
+    QString name;
+};
+
+class VirtualMachinePrivate;
+class SFDK_EXPORT VirtualMachine : public QObject
+{
+    Q_OBJECT
+
+public:
+    class ConnectionUi;
+
+    // Keep in sync with VmConnection::str()
+    enum State {
+        Disconnected,
+        Starting,
+        Connecting,
+        Error,
+        Connected,
+        Disconnecting,
+        Closing
+    };
+    Q_ENUM(State)
+
+    enum ConnectOption {
+        NoConnectOption = 0x00,
+        AskStartVm = 0x01,
+        Block = 0x02,
+    };
+    Q_DECLARE_FLAGS(ConnectOptions, ConnectOption)
+
+    enum Synchronization {
+        Asynchronous,
+        Synchronous
+    };
+
+    ~VirtualMachine() override;
+
+    QUrl uri() const;
+    QString type() const;
+    QString displayType() const;
+    // FIXME add qualifiedName() that would include type information?
+    QString name() const;
+
+    State state() const;
+    QString errorString() const;
+
+    QSsh::SshConnectionParameters sshParameters() const; // FIXME internal
+    void setSshParameters(const QSsh::SshConnectionParameters &sshParameters);
+
+    bool isHeadless() const;
+    void setHeadless(bool headless);
+
+    bool isAutoConnectEnabled() const; // FIXME internal?
+    void setAutoConnectEnabled(bool autoConnectEnabled);
+
+    bool isOff(bool *runningHeadless = 0, bool *startedOutside = 0) const;
+    bool lockDown(bool lockDown);
+    bool isLockedDown() const;
+
+    int memorySizeMb() const;
+    void setMemorySizeMb(int memorySizeMb, const QObject *context,
+            const Functor<bool> &functor);
+    static int availableMemorySizeMb();
+
+    int cpuCount() const;
+    void setCpuCount(int cpuCount, const QObject *context,
+            const Functor<bool> &functor);
+    static int availableCpuCount();
+
+    // FIXME rename as storageCapacityMb
+    int vdiCapacityMb() const;
+    void setVdiCapacityMb(int vdiCapacityMb, const QObject *context,
+            const Functor<bool> &functor);
+
+    bool hasPortForwarding(quint16 hostPort, QString *ruleName = nullptr) const;
+    void addPortForwarding(const QString &ruleName, const QString &protocol,
+            quint16 hostPort, quint16 emulatorVmPort, const QObject *context,
+            const Functor<bool> &functor);
+    void removePortForwarding(const QString &ruleName, const QObject *context,
+            const Functor<bool> &functor);
+
+    QStringList snapshots() const;
+
+    void refreshConfiguration(const QObject *context, const Functor<bool> &functor);
+    // FIXME API style?
```

Comment:
Suspicious comment. Is this something you intended to fix before merging?

Reply:
Not planned to be fixed anytime soon. The same applies to all the other TODO and FIXME comments left there. They highlight the ideas for further development with no schedule set.